astropy/wcs/src/wcslib_wrap.c

```diff
@@ -157,6 +159,145 @@ convert_rejections_to_warnings() {
   return status;
 }
 
+
+/***************************************************************************
+ * wtbarr-related global variables and functions                           *
+ ***************************************************************************/
+
+static PyObject *get_wtbarr_data = NULL;
+
+
+void _set_wtbarr_callback(PyObject* callback) {
+  Py_XINCREF(callback);         /* Add a reference to new callback */
+  Py_XDECREF(get_wtbarr_data);  /* Dispose of previous callback */
+  get_wtbarr_data = callback;   /* Remember new callback */
+}
+
+
+int _update_wtbarr_from_hdulist(PyObject *hdulist, struct wtbarr *wtb) {
+  PyArrayObject *arrayp=NULL;
+  PyObject *result=NULL, *arglist = NULL;
+  int i, naxis, nelem, *naxes;
+  npy_intp *npy_naxes;
+  npy_double *appayp_data;
+
+  if (hdulist == NULL || hdulist == Py_None) {
+    PyErr_SetString(PyExc_ValueError,
+                    "HDUList is required to retrieve -TAB coordinates "
+                    "and/or indices.");
+    return 0;
+  }
+
+  if (wtb->ndim < 1) {
+    PyErr_SetString(PyExc_ValueError, "Number of dimensions should be positive.");
+    return 0;
+  }
+
+  result = PyObject_CallFunction(get_wtbarr_data, "(OsiiCsii)", hdulist,
+      wtb->extnam, wtb->extver, wtb->extlev, wtb->kind, wtb->ttype, wtb->row,
+      wtb->ndim);
+      
+  if (result == NULL) return 0;
+
+  arrayp = (PyArrayObject *)PyArray_FromAny(result,
+      PyArray_DescrFromType(NPY_DOUBLE), 0, 0, NPY_ARRAY_CARRAY, NULL);
+
+  if (arrayp == NULL) {
+    PyErr_SetString(PyExc_TypeError, "Unable to convert wtbarr callback "
+                    "result to a numpy.ndarray.");
+    Py_DECREF(result);
+    return 0;
+  }
+
+  if ((PyObject *)arrayp != result) Py_DECREF(result);
```

Comment:
I think you can just unconditionally `Py_DECREF(result)` after the `PyArray_FromAny` call. It seems like it always returns a new reference and doesn't steal a reference for `result`. Or was that something that didn't work?

Reply:
If `result` is already a C contiguous array of doubles then `arrayp` is `result`. I do not think the ref counter is incremented. I would rather leave this as is.

Reply:
> I would rather leave this as is.

The problem is that failing to `DECREF` an owned reference will result in a memory-leak, but `DECREF` a borrowed reference will likely run into a segmentation fault. So I think we need to figure this out - if you don't have the time I can do that later today.

I'm almost sure that it increments the reference counter even if it's already the correct array (unfortunately I'm currently not on my computer so I cannot check). The routine creating the array should be [`PyArray_GetArrayParamsFromObject`](https://github.com/numpy/numpy/blob/v1.17.4/numpy/core/src/multiarray/ctors.c#L1600).

Reply:
Good catch! Indeed, it increments reference count even when input is passed through. I'll fix that.